File: geopdes_maxwell/inst/operators/op_v_gradp.m
% OP_V_GRADP: assemble the matrix B = [b(i,j)], b(i,j) = (epsilon grad p_i, v_j).
%
%   mat = op_v_gradp (spv, spp, msh, epsilon);
%   [rows, cols, values] = op_v_gradp (spv, spp, msh, epsilon);
%
% INPUT:
%    
%   spv:     structure representing the space of vectorial trial functions  (see sp_vector_2d_curl_transform/sp_evaluate_col)
%   spp:     structure representing the space of scalar test functions (see sp_bspline_2d/sp_evaluate_col)
%   msh:     structure containing the domain partition and the quadrature rule (see msh_2d/msh_evaluate_col)
%   epsilon: physical parameter
%
% OUTPUT:
%
%   mat:    assembled matrix
%   rows:   row indices of the nonzero entries
%   cols:   column indices of the nonzero entries
%   values: values of the nonzero entries
% 
% Copyright (C) 2009, 2010 Carlo de Falco
% Copyright (C) 2011 Rafael Vazquez
%
%    This program is free software: you can redistribute it and/or modify
%    it under the terms of the GNU General Public License as published by
%    the Free Software Foundation, either version 3 of the License, or
%    (at your option) any later version.

%    This program is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU General Public License for more details.
%
%    You should have received a copy of the GNU General Public License
%    along with this program.  If not, see <http://www.gnu.org/licenses/>.

function varargout = op_v_gradp (spv, spp, msh, coeff)

  ndir = size (spp.shape_function_gradients, 1);

  rows = zeros (msh.nel * spp.nsh_max * spv.nsh_max, 1);
  cols = zeros (msh.nel * spp.nsh_max * spv.nsh_max, 1);
  values = zeros (msh.nel * spp.nsh_max * spv.nsh_max, 1);

  ncounter = 0;
  for iel = 1:msh.nel
    if (all (msh.jacdet(:, iel)))
      jacdet_weights = reshape (msh.jacdet(:,iel) .* ...
                         msh.quad_weights(:, iel) .* coeff(:,iel), 1, msh.nqn);

      gradp_iel = reshape (spp.shape_function_gradients(:, :, 1:spp.nsh(iel), iel), ...
                                     ndir, msh.nqn, spp.nsh(iel));
      shpv_iel = reshape (spv.shape_functions(:, :, 1:spv.nsh(iel), iel), ...
                                     ndir, msh.nqn, spv.nsh(iel));

      gradp_times_jw = bsxfun (@times, jacdet_weights, gradp_iel);
      for idof = 1:spp.nsh(iel)
        rows(ncounter+(1:spv.nsh(iel))) = spp.connectivity(idof, iel);
        cols(ncounter+(1:spv.nsh(iel))) = spv.connectivity(1:spv.nsh(iel), iel);

        aux_val = bsxfun (@times, gradp_times_jw(:,:,idof), shpv_iel);
        values(ncounter+(1:spv.nsh(iel))) = sum (sum (aux_val, 2), 1);
        ncounter = ncounter + spv.nsh(iel);
      end
    else
      warning ('geopdes:jacdet_zero_at_quad_node', 'op_v_gradp: singular map in element number %d', iel)
    end
  end

  if (nargout == 1)
    varargout{1} = sparse (rows(1:ncounter), cols(1:ncounter), ...
                           values(1:ncounter), spp.ndof, spv.ndof);
  elseif (nargout == 3)
    varargout{1} = rows(1:ncounter);
    varargout{2} = cols(1:ncounter);
    varargout{3} = values(1:ncounter);
  else
    error ('op_v_gradp: wrong number of output arguments')
  end

end


%% COPY OF THE FIRST VERSION OF THE FUNCTION (MORE UNDERSTANDABLE)
% function mat = op_v_gradp (spv, spp, msh, coeff)
% 
%   mat = spalloc(spp.ndof, spv.ndof, 1);
%   ndir = size (spp.shape_function_gradients, 1);
%   for iel = 1:msh.nel
%     if (all (msh.jacdet(:,iel)))
%       mat_loc = zeros (spp.nsh(iel), spv.nsh(iel));
%       for idof = 1:spp.nsh(iel)
%         ishg = reshape(spp.shape_function_gradients(:,:,idof,iel),ndir,[]);
%         for jdof = 1:spv.nsh(iel)
%           jshg = reshape(spv.shape_functions(:,:,jdof,iel),ndir,[]);
% % The cycle on the quadrature points is vectorized
%           %for inode = 1:msh.nqn
%             mat_loc(idof, jdof) = mat_loc(idof, jdof) + ...
%              sum (msh.jacdet(:, iel) .* msh.quad_weights(:, iel) .* ...
%              sum (ishg .* jshg, 1).' .* coeff(:, iel));
%           %end  
%         end
%       end
%       mat(spp.connectivity(:, iel), spv.connectivity(:, iel)) = ...
%         mat(spp.connectivity(:, iel), spv.connectivity(:, iel)) + mat_loc;
%     else
%       warning ('geopdes:jacdet_zero_at_quad_node', 'op_v_gradp: singular map in element number %d', iel)
%     end
%   end
% 
% end

File: tags/release_2_0_3/geopdes_maxwell/src/op_v_gradp.cc
/* Copyright (C) 2009, 2010, 2011 Carlo de Falco
   Copyright (C) 2011 Rafael Vazquez

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
 
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#include "geopdes.h"

DEFUN_DLD(op_v_gradp, args, nargout,"\n\
OP_V_GRADP: assemble the matrix B = [b(i,j)], b(i,j) = (epsilon grad p_i, v_j). \n\
\n\
   mat = op_v_gradp (spv, spp, msh, epsilon);\n\
\n\
 INPUT:\n\
\n\
   spv:     structure representing the space of vectorial trial functions  (see sp_bsp_hcurl_2d_phys). \n\
   spp:     structure representing the space of scalar test functions (see sp_bspline_2d_phys). \n\
   msh:     structure containing the domain partition and the quadrature rule (see msh_push_forward_2d). \n\
   epsilon: physical parameter. \n\
\n\
 OUTPUT: \n\
\n\
   mat: assembled matrix. \n\
")
{
  
  octave_value_list retval;

  geopdes_mesh  msh (args(2).map_value ());
  geopdes_space spv (args(0).map_value (), msh);
  geopdes_space spp (args(1).map_value (), msh);
  Matrix coeff = args(3).matrix_value();

  SparseMatrix mat;

  if (!error_state)
    {

      const octave_idx_type nel = msh.nel (), nqn = msh.nqn (), ncomp = spv.ncomp (), ndof_spp = spp.ndof (), 
        nsh_max_spp = spp.nsh_max (), ndof_spv = spv.ndof (), nsh_max_spv = spv.nsh_max ();

#if OCTAVE_API_VERSION_NUMBER>37
      dim_vector dims (nel * nsh_max_spv * nsh_max_spp, 1);
      Array <octave_idx_type> I (dims, 0);
      octave_idx_type* Iptr = I.fortran_vec ();

      Array <octave_idx_type> J (dims, 0);
      octave_idx_type* Jptr = J.fortran_vec ();

      Array <double> V (dims, 0.0);    
      double* Vptr = V.fortran_vec ();
#else
      ColumnVector I (nel * nsh_max_spv * nsh_max_spp, 0);
      double* Iptr = I.fortran_vec ();

      ColumnVector J (nel * nsh_max_spv * nsh_max_spp, 0);
      double* Jptr = J.fortran_vec ();

      ColumnVector V (nel * nsh_max_spv * nsh_max_spp, 0.0);
      double* Vptr = V.fortran_vec ();
#endif

      octave_idx_type counter = 0, iel, inode, idof, jdof, icmp;
 
      {      
        for ( iel=0; iel < nel; iel++) 
          if (msh.area (iel) > 0.0)
            {
              const octave_idx_type nsh_p = spp.nsh (iel);
              const octave_idx_type nsh_v = spv.nsh (iel);
              double jacdet_weights[nqn];

              for ( inode = 0; inode < nqn; inode++)
                {
                  jacdet_weights[inode] = msh.jacdet (inode, iel) *
                    msh.weights (inode, iel) * coeff (inode, iel);
                }

              double shpv[nsh_v][nqn][ncomp];
              double shgp[nsh_p][nqn][ncomp];
              octave_idx_type conn_v[nsh_max_spv];
              octave_idx_type conn_p[nsh_max_spp];

              for ( idof = 0; idof < nsh_p; idof++) 
                for ( inode = 0; inode < nqn; inode++)
                  for ( icmp = 0; icmp < ncomp; icmp++)
                    shgp[idof][inode][icmp] = spp.shape_function_gradients (0, icmp, inode, idof, iel);

              for ( jdof = 0; jdof < nsh_v; jdof++) 
                for ( inode = 0; inode < nqn; inode++)
                  for ( icmp = 0; icmp < ncomp; icmp++)
                    shpv[jdof][inode][icmp] = spv.shape_functions (icmp, inode, jdof, iel);

              // Cache element data to speed-up memory access
              spp.cache_element_connectivity (iel, (octave_idx_type*)conn_p);
              spv.cache_element_connectivity (iel, (octave_idx_type*)conn_v);


              for ( idof = 0; idof < nsh_p; idof++) 
                {
                  for ( jdof = 0; jdof < nsh_v; jdof++)
                    {
                      counter = jdof + nsh_v * (idof + nsh_p * iel);

                      Iptr[counter] = conn_p[idof] - 1;
                      Jptr[counter] = conn_v[jdof] - 1;
                      Vptr[counter] = 0.0;
                      for ( inode = 0; inode < nqn; inode++)
                        {
                          if (msh.weights (inode, iel) > 0.0)
                            {			
                              double s = 0.0;
                              for ( icmp = 0; icmp < ncomp; icmp++)
                                s += shgp[idof][inode][icmp] * shpv[jdof][inode][icmp];
                              Vptr[counter] += jacdet_weights[inode] * s;
                            }  
                        } // end for inode		  
                      counter++;
                    } // end for jdof
                } // end for idof
            } else {
            {warning_with_id ("geopdes:zero_measure_element", "op_v_gradp: element %d has 0 area (or volume)", iel);}
          }  // end for iel, if area > 0
      } // end of parallel section

      if (nargout == 1) 
        {
          mat = SparseMatrix (V, I, J, ndof_spp, ndof_spv, true);
          retval(0) = octave_value (mat);
        } 
      else if (nargout == 3)
	{
          for ( icmp = 0; icmp <= counter; icmp++) 
            {
              Iptr[icmp] += 1;
              Jptr[icmp] += 1;
            }
          retval(0) = octave_value (I);
          retval(1) = octave_value (J);
          retval(2) = octave_value (V);
        }

    } // end if !error_state
  return retval;
}
